traits/trait_types.py

```diff
@@ -1700,127 +1701,263 @@ def __init__(
 
+# Machinery to infer the value type to use for a Range trait.
+
+
+def _is_numeric(value):
+    """
+    Determine whether a value is numeric, and return its type if so.
+
+    Here numeric means either int-like or float-like.
+
+    Parameters
+    ----------
+    value : object
+        The value to be tested.
+
+    Returns
+    -------
+    is_numeric : bool
+        True if the value has a numeric type, else False
+    value_type : one of (int, float, None)
+        int if the value is integer-like; float if the value is float-like,
+        None if neither of those is true.
+
+    Raises
+    ------
+    TraitError
+        If *value* is not None, not a string, not an integer and
+        not a floating-point number.
+    """
+    if value is None:
+        return False, None
+
+    try:
+        _validate_int(value)
+    except TypeError:
+        pass
+    else:
+        return True, int
+
+    try:
+        _validate_float(value)
+    except TypeError:
+        pass
+    else:
+        return True, float
+
+    raise TraitError(
+        "value should be an integer, a floating-point number, or a string. "
+        "Got {!r}".format(value)
+    )
+
+
+def _infer_range_value_trait(low, high, default):
+    """
+    Infer the value type of a Range trait from its parameters.
+
+    Parameters
+    ----------
+    low, high, default : object
+        The low bound, high bound and default value given at
+        Range trait definition time.
+
+    Returns
+    -------
+    value_trait : TraitType
+        Value type for the range: an instance of either Int or Float.
+
+    Raises
+    ------
+    TraitError
+        If any of *low*, *high* and *default* is not None, is not a string,
+        not an integer and not a floating-point number.
+
+    """
+    value_types = [
+        value_type
+        for is_numeric, value_type in map(_is_numeric, [low, high])
+        if is_numeric
+    ]
+
+    # For backwards compatibility, assume Float if no value_type is given.
+    if not value_types:
+        warnings.warn(
+            message=(
+                "Unable to infer a value type for this Range instance. "
+                "Using Float(). This warning can be silenced by passing "
+                "value_trait=Int() or value_trait=Float() as appropriate "
+                "at Range creation time. This warning may become an error "
+                "in a future version of Traits. "
+            ),
+            category=DeprecationWarning,
+            stacklevel=3,
+        )
+        trait_type = Float
+    else:
+        trait_type = Float if float in value_types else Int
+
+    return trait_type() if default is None else trait_type(default)
+
 
 class BaseRange(TraitType):
-    """ Defines a trait whose numeric value must be in a specified range.
+    """ A trait whose value must be within a specified range.
     """
 
     def __init__(
         self,
         low=None,
         high=None,
         value=None,
+        low_name=None,
```

Comment:
Now that the `Range` trait can use arbitrary types, there may be a case for using a default other than `None` for `low`, `high` and `value`, just in case somehow a value of `None` is usable.

Reply:
I think using `None` a part of a custom ordering in a way that is significant is sufficiently weird that we don't need to worry that if someone does that they might need to subclass to get the validation right.

Reply:
Agreed; let's leave it